aten/src/ATen/native/MaxUnpooling.cpp

```diff
@@ -0,0 +1,627 @@
+      scalar_t* output_p_k = rawOutput + finalOutputOffset;
+      scalar_t* input_p_k = rawInput + finalInputOffset;
+      int64_t* ind_p_k = rawIndices + finalInputOffset;
+
+      int64_t maxp;
+      for (int64_t i = 0; i < inputHeight; i++) {
+        for (int64_t j = 0; j < inputWidth; j++) {
+          maxp = ind_p_k[i * inputWidth + j];
+          if (maxp < 0 || maxp >= outputWidth * outputHeight) {
+#pragma omp critical
+            {
+              has_error = true;
+              error_index = maxp;
+            }
+          } else {
+            output_p_k[maxp] = input_p_k[i * inputWidth + j];
+          }
+        }
+      }
+    }
+    if (has_error) {
+      AT_ERROR(
+          "Found an invalid max index",
+          error_index,
+          " (output volumes are of size ",
+          outputHeight,
+          "x",
+          outputWidth);
+    }
+  }
+  return output;
+}
+
+Tensor& max_unpooling2d_forward_out_cpu(
+    Tensor& output,
+    const Tensor& self,
+    const Tensor& indices,
+    IntList output_size) {
+  AT_CHECK(
+      indices.scalar_type() == at::ScalarType::Long,
+      "elements in indices should be type Long");
+  AT_CHECK(
+      output_size.size() == 2,
+      "There should be exactly two elements (height, width) in output_size");
+  AT_CHECK(
+      (self.ndimension() == 3 || self.ndimension() == 4),
+      "Input to max_unpooling2d should be a 3d or 4d Tensor");
+  AT_CHECK(
+      self.sizes() == indices.sizes(),
+      "Shape of indices should match shape of input");
+
+  // is_empty check
+  AT_CHECK(self.numel() > 0, "Input must be non-empty");
+
+  auto outputHeight = output_size[0];
+  auto outputWidth = output_size[1];
+
+  int64_t numBatch = 1;
+  int64_t numChannels;
+  if (self.ndimension() == 4) {
+    numBatch = self.size(0);
+    numChannels = self.size(1);
+  } else {
+    numChannels = self.size(0);
+  }
+
+  auto self_contiguous = self.contiguous();
+  auto indices_contiguous = indices.contiguous();
+
+  if (self_contiguous.ndimension() == 3) {
+    output.resize_({numChannels, outputHeight, outputWidth});
+  } else {
+    output.resize_({numBatch, numChannels, outputHeight, outputWidth});
+  }
+  output.zero_();
+
+  AT_DISPATCH_FLOATING_TYPES(
+      self.type(), "max_unpooling2d_forward_out_cpu_frame", ([&] {
+        max_unpooling2d_forward_out_cpu_frame<scalar_t>(
+            output,
+            self_contiguous,
+            indices_contiguous,
+            output_size[0],
+            output_size[1]);
+      }));
+  return output;
+};
+
+Tensor max_unpooling2d_forward_cpu(
+    const Tensor& self,
+    const Tensor& indices,
+    IntList output_size) {
+  AT_CHECK(
+      (self.ndimension() == 3 || self.ndimension() == 4),
+      "Input to max_unpooling2d should be a 4d or 5d Tensor",
+      self.sizes());
+  AT_CHECK(
+      output_size.size() == 2,
+      "There should be exactly two elements (height, width) in output_size");
+
+  auto output = at::empty(
```

Comment:
would it work to do something like `auto output = at::empty({0}, self.options())` here?  Because how you've done it here requires additional checks and the shape isn't even necessarily correct (output can be 3d, but you always set it to 4d here).